### PR TITLE
feat: update dharma.js instantiators

### PR DESCRIPTION
**Pull request by chrismin**

### CODE REVIEW

**Comment by saturnial** [comment]
feels like we should discuss as a team what the ideal API here is.

two things i'd object to in this PR:
- we lost the simpler implementation of just passing in the blockchain host as a string
- seems pretty wild to throw based on web3 version -- could see a lot of ppl getting confused. feels like it shouldn't be the default option.

@nadavhollander seems to have strong opinions though. would love for him to weigh in.

**Comment by nadavhollander** [comment]
I feel very strongly that users should be able to pass `web3` in.  Instantiating only with host string makes an extremely narrow assumption on what users are using `dharma.js` for -- i.e. to connect w/ an RPC node via HTTP where account 0 is permanently unlocked.

There is a much, much wider variety of ways `web3` can be used, and those use-cases aren't necessarily esoteric or niche; many are extremely common.  To list some examples...

1. Caching functionality
2. Websocket Connectivity (necessary for using Infura AFAIK)
3. Fallback Nodes
4. JS Instantiated Wallet (i.e. plaintext key imports, very useful for working on a server / with Infura)
5. Hardware Wallets
6. Custom Nonce Resets
7. Dynamic Gas Pricing

Even if we assume 90% of people don't care about the above and only care about Localhost dapps / MetaMask dapps, there's _still_ a lot of reasons why a user would care about accessing the raw `web3` instance.  As an great example, dApps care about detecting whether web3 is present, unlocked, and switched onto the correct network.  If `web3` is tucked away / abstracted into Dharma, figuring out how to do this is highly non-intuitive.

At ETHIndia, this was IMO one of the biggest issues faced -- I had to show 2 different teams how to hack our APIs and inject web3 in.  I think the simplicity gain here is narrow to a very specific use-case, and is greatly outweighed by the ways in which we're hurt by inflexibility here.

**Comment by chrismin** [comment]
will add tests if and when we come to a consensus on the code itself!

**Comment by nadavhollander** [comment]
I still feel pretty strongly that this is an odd pattern that causes many more problems than it solves.  I think it will make our documentation more confusing (i.e. there will have to be a `Setting a Custom Web3 Provider` section, which is much less clean than just encapsulating everything in `Getting Started`).  Additionally, we're making an opinionated choice on the default way people should set up `web3` -- which is OK in principle -- but the "opinion" we're advocating is a very narrow slice of the ways people use `web3` instances that reflects our internal workflow more than it does how people in _general_ use `web3`.

I won't die on this hill and I'm happy to try it this way and see how it goes in ETHBerlin -- but I will flag that I think this will put people who are building something that's more than a localhost PoC in a highly error-prone position.

**Comment by chrismin** [comment]
Hmm... I think in that case the looming question of supporting web3 `1.0.0` becomes more relevant, as it's a pretty poor experience and a high barrier to entry to throw on instantiating with a web3 > `0.2.x`.  @nadavhollander  do you have a sense of how widely adopted web3 `1.0.0` is?  I understand it's still in beta, but there were definitely some teams using it at ETHIndia.  Also, do you have a sense of how much work it would be to update dharma.js to use `1.0.0`, or, even better, make it version agnostic?

I think if we can reasonably solve or work around this issue, passing in web3 upon instantiation of dharma.js makes sense.  I think it makes sense to think of web3 as a dependency, especially since, generally speaking, dharma.js may only comprise a single portion of a dApp's interactions with the blockchain, and the ground dharma.js covers isn't a superset of the ground web3 covers and vice versa.

**Comment by nadavhollander** [comment]
> I think it makes sense to think of web3 as a dependency, especially since, generally speaking, dharma.js may only comprise a single portion of a dApp's interactions with the blockchain, and the ground dharma.js covers isn't a superset of the ground web3 covers and vice versa.

Yup, exactly where my head is at.

> Hmm... I think in that case the looming question of supporting web3 1.0.0 becomes more relevant, as it's a pretty poor experience and a high barrier to entry to throw on instantiating with a web3 > 0.2.x. @nadavhollander do you have a sense of how widely adopted web3 1.0.0 is? I understand it's still in beta, but there were definitely some teams using it at ETHIndia. Also, do you have a sense of how much work it would be to update dharma.js to use 1.0.0, or, even better, make it version agnostic?

Oh man -- the web3 1.0.0 divide is a shitshow.  My impression (which may be false, could be a good idea to put out a survey on this) is that most use < 1.0.0 because it's the best documented and is not in beta.

I think making `dharma.js` compatible with both is a medium-scope project for sure (at least a Medium on our product planning scale), though possible.  I also don't think the versioning problem goes away if we provide the option of setting a custom `web3` instance outside of the constructor -- therefore we kinda need to do _something_ anyways, and, if we're not throwing when they set the web3 instance, they're going to get mystery errors elsewhere which will likely be more confusing.

**Comment by chrismin** [comment]
> I also don't think the versioning problem goes away if we provide the option of setting a custom web3 instance outside of the constructor

Right--but it does move the potential error to outside of the constructor, which reduces friction to getting started.

Having said that, if we're trying to fail fast, perhaps we want that error to be in the constructor... Passing in an instance of web3 when initializing Dharma.js seems to be a relatively advanced maneuver , so perhaps it makes sense to fail fast in this case.

File: src/dharma.ts [hunk 1]
```diff
@@ -16,6 +16,20 @@ import * as DharmaTypes from "./types";
 class Dharma {
     public static Types = DharmaTypes;
 
+    public static withBlockchainNode(blockchainNode: string) {
```
Comment (saturnial) [inline src/dharma.ts]: i would write this signature as:

`public static initialize(host: string): Dharma`

File: src/dharma.ts [hunk 2]
```diff
@@ -16,6 +16,20 @@ import * as DharmaTypes from "./types";
...
+        const web3Provider = new Web3.providers.HttpProvider(blockchainNode);
+        const web3 = new Web3(web3Provider);
+
+        // Verify that the node is reachable
+        web3.version.getNode((error, result) => {
```
Comment (saturnial) [inline src/dharma.ts]: this operation is async within the context of a non-async fn.

File: src/dharma.ts [hunk 3]
```diff
@@ -27,25 +41,27 @@ class Dharma {
 
     public readonly web3: Web3;
 
-    constructor(blockchainHost?: string, addressBook: DharmaTypes.AddressBook = {}) {
+    constructor(web3?: Web3, addressBook: DharmaTypes.AddressBook = {}) {
```
Comment (saturnial) [inline src/dharma.ts]: there is a world where this web3 instance is not optional.

`Dharma.default()` -> instantiates Dharma by pulling web3 off the window.

File: src/dharma.ts [hunk 4]
```diff
@@ -27,25 +41,27 @@ class Dharma {
...
             typeof (window as any).web3 !== "undefined"
         ) {
-            // If web3 is available via the browser window, instantiate web3 via the current provider.
+            // If Web3 is available via the browser window, instantiate Web3 via the current provider.
             this.web3 = new Web3((window as any).web3.currentProvider);
         } else {
             // Otherwise throw...
-            throw new Error("Pass in the address of your blockchain node.");
+            throw new Error(
+                "A Web3 instance was not specified as a parameter and Web3 was not found on the window.",
```
Comment (saturnial) [inline src/dharma.ts]: I would consider adding a constant at the top that specifies all the errors thrown.

File: __test__/unit/dharma.spec.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                test("successfully instantiates Dharma.js", () => {
+                    const dharma = Dharma.withBlockchainHost("http://localhost:8545");
+
+                    expect(dharma).toBeInstanceOf(Dharma);
+                });
+            });
+
+            describe("when incorrectly formatted blockchain host is passed in", () => {
+                test("throws", () => {
+                    expect(() => Dharma.withBlockchainHost("a random string")).toThrow();
```
Comment (saturnial) [inline __test__/unit/dharma.spec.ts]: i don't see a static method with this name

File: src/dharma.ts [hunk 6]
```diff
@@ -58,6 +74,24 @@ class Dharma {
         this.blockchain = new BlockchainAPI(this.web3, this.contracts);
         this.logs = new LogsAPI(this.web3, this.contracts);
     }
+
+    private verifyWeb3Version(web3: Web3) {
```
Comment (saturnial) [inline src/dharma.ts]: this function smells funky to me. is our version requirements so strict that it needs to be the exact version? and do we not have a minimum version and max version that we can just specify as constants somewhere.

File: src/dharma.ts [hunk 7]
```diff
@@ -13,19 +14,39 @@ import {
...
+        ) {
+            const web3Provider = (window as any).web3.currentProvider;
+            const web3 = new Web3(web3Provider);
+            return new Dharma(web3);
+        } else {
+            throw new Error(DharmaInstantiationErrors.WEB3_NOT_FOUND_ON_WINDOW);
+        }
+    }
+
+    public static initializeWithBlockchainNode(blockchainNode: string): Dharma {
```
Comment (saturnial) [inline src/dharma.ts]: `initializeWithNode(host: string): Dharma`

**Comment by graemecode** [comment]
Just through my 2 minute perusal of this, one core problem I think I have is as follows (@chrismin @nadavhollander @saturnial )


If I want to use the global web3 from `window.web3` (e.g. via MetaMask, relayer kit) then I use `Dharma.initialize()`. But if I'm developing locally and want to connect to local node, then I would probably be instructed to use `new Dharma();` So the implication is that I should have some conditional:

```javascript
if (not using MetaMask) {
  dharma = Dharma.initialize();
} else {
  dharma = new Dharma(myWeb3Instance);
}
```

But I imagine this is quite confusing, so I imagine the appropriate thing is to always pass in web3, a la:

```
const dharma = new Dharma(myWeb3Version);
```

But this will throw with the default web3 that gets added by `yarn add web3`. So there needs to be specific instruction that you need to use the "non-default" web3, and that the rest of your app to the extent that it uses web3 needs to use that version as well.

So here's my problem with this PR:

1. There is an unnecessary function `initialize` that does something that's very useful (I.E. it allows you to use a global web3 very easily. 

Note that this could work for all scenarios as follows:
```javascript
window.web3 = new Web3(myProvider):
const dharma = new Dharma(); // default here to global web3, whether I pass it in or metamask, etc.
```

But instead there are two ways of constructing Dharma. Which is confusing.

2. The default web3 install is going to throw on the canonical way of instantiation!

**Pull request closed without merging**